### PR TITLE
[fix] #2416: Fix lints on macOS arm

### LINKED ISSUE #2416: Ensure that M1 Mac `clippy::variant_size_differences` is fixed

### PR DESCRIPTION
**Pull request by ales-tsurko**
Signed-off-by: Ales Tsurko <ales.tsurko@gmail.com>

### Description of the Change

A cross-platform wrapper for `BlockValue` added to encapsulate platform dependent code. On macOS with arm architecture `Value::Block` inner is now boxed. 

### Issue

Closes #2416 

### Benefits

No more linter errors on M1.

### Possible Drawbacks

As it's now a smart pointer, it requires dereferencing, when the inner value is needed.

### CODE REVIEW

**Comment by Erigara** [comment]
Your solution has conflicts with @Arjentix recent PR #2436.

File: data_model/src/lib.rs [hunk 1]
```diff
@@ -338,7 +343,81 @@ pub enum Value {
     /// [`struct@Hash`]
     Hash(Hash),
     /// Block
-    Block(BlockValue),
+    Block(BlockValueWrapper),
+}
+
+/// Cross-platform wrapper for `BlockValue`.
```
Comment (Erigara) [inline data_model/src/lib.rs:349]: IMO we introduce unnecessary complexity here by make platform specific implementation of `BlockValueWrapper`.

I think maybe just `Box<BlockValue>` could work?

File: data_model/src/lib.rs [hunk 2]
```diff
@@ -338,7 +343,81 @@ pub enum Value {
...
+#[cfg(not(all(target_arch = "aarch64", target_os = "macos")))]
+#[derive(
+    Debug, Clone, PartialEq, Eq, Decode, Encode, Deserialize, Serialize, IntoSchema, PartialOrd, Ord,
+)]
```
Comment (Erigara) [inline data_model/src/lib.rs:353]: I believe [`#[serde(transparent)]`](https://serde.rs/container-attrs.html#transparent) should be used here.

File: data_model/src/lib.rs [hunk 3]
```diff
@@ -338,7 +343,81 @@ pub enum Value {
...
+
+#[cfg(not(all(target_arch = "aarch64", target_os = "macos")))]
+impl AsRef<BlockValue> for BlockValueWrapper {
+    fn as_ref(&self) -> &BlockValue {
+        &self.0
+    }
+}
+
+#[cfg(all(target_arch = "aarch64", target_os = "macos"))]
+impl Deref for BlockValueWrapper {
```
Comment (Erigara) [inline data_model/src/lib.rs:380]: Check `derive_more::Deref`.

File: data_model/src/lib.rs [hunk 4]
```diff
@@ -338,7 +343,81 @@ pub enum Value {
...
+
+#[cfg(not(all(target_arch = "aarch64", target_os = "macos")))]
+impl From<BlockValue> for BlockValueWrapper {
+    fn from(block_value: BlockValue) -> Self {
+        BlockValueWrapper(block_value)
+    }
+}
+
+#[cfg(all(target_arch = "aarch64", target_os = "macos"))]
+impl From<BlockValue> for BlockValueWrapper {
```
Comment (Erigara) [inline data_model/src/lib.rs:403]: Check `derive_more::From` and `derive_more::Into`.

File: data_model/src/lib.rs [hunk 5]
```diff
@@ -338,7 +343,81 @@ pub enum Value {
...
+impl Deref for BlockValueWrapper {
+    type Target = BlockValue;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+#[cfg(not(all(target_arch = "aarch64", target_os = "macos")))]
+impl AsRef<BlockValue> for BlockValueWrapper {
```
Comment (Erigara) [inline data_model/src/lib.rs:373]: Check `derive_more::AsRef`.

File: data_model/src/lib.rs (hunk 1 above)
Comment (ales-tsurko) [reply data_model/src/lib.rs:349]: Please, read the requirements

File: data_model/src/lib.rs [hunk 6]
```diff
@@ -338,7 +343,81 @@ pub enum Value {
...
+impl Deref for BlockValueWrapper {
+    type Target = BlockValue;
+
+    fn deref(&self) -> &Self::Target {
+        self.0.as_ref()
+    }
+}
+
+#[cfg(all(target_arch = "aarch64", target_os = "macos"))]
+impl AsRef<BlockValue> for BlockValueWrapper {
```
Comment (appetrosyan) [inline data_model/src/lib.rs:389]: Derive

File: data_model/src/lib.rs [hunk 7]
```diff
@@ -338,7 +343,81 @@ pub enum Value {
...
+    fn as_ref(&self) -> &BlockValue {
+        self.0.as_ref()
+    }
+}
+
+#[cfg(not(all(target_arch = "aarch64", target_os = "macos")))]
+impl From<BlockValue> for BlockValueWrapper {
```
Comment (appetrosyan) [inline data_model/src/lib.rs:396]: Derive

File: data_model/src/lib.rs [hunk 8]
```diff
@@ -391,6 +470,12 @@ impl Value {
     }
 }
 
+impl From<BlockValue> for Value {
```
Comment (appetrosyan) [inline data_model/src/lib.rs:473]: Consider `derive_more::From` with a type annotation.

File: data_model/src/lib.rs [hunk 9]
```diff
@@ -600,6 +685,18 @@ where
     }
 }
 
+impl TryFrom<Value> for BlockValue {
```
Comment (appetrosyan) [inline data_model/src/lib.rs:688]: Consider deriving.

Comment (ales-tsurko) [reply data_model/src/lib.rs:662]: There's no `TryFrom`, and for `BlockValue` we can't use `TtryInto` as it has too many fields.

File: data_model/src/lib.rs (hunk 8 above)
Comment (appetrosyan) [reply data_model/src/lib.rs:449]: `FromVariant` is redundant. But I agree, that we can remove it in one fell swoop later.

**Review by appetrosyan — APPROVED** [review]

File: data_model/src/lib.rs [hunk 10]
```diff
@@ -341,7 +343,60 @@ pub enum Value {
...
+    Clone,
+    Debug,
+    Decode,
+    Deref,
+    Deserialize,
+    Encode,
+    Eq,
+    From,
+    Into,
+    IntoSchema,
```
Comment (Erigara) [inline data_model/src/lib.rs:362]: Shouldn't we manually implement `IntoSchema` to hide the presence of the `BlockValueWrapper` from the generated schema?
It's seems like our inner stuff.

**Review by Erigara — APPROVED** [review]

File: data_model/src/lib.rs (hunk 10 above)
Comment (ales-tsurko) [reply data_model/src/lib.rs:362]: Oh. A good point. I'll take a look. Thanks!

Comment (appetrosyan) [reply data_model/src/lib.rs:362]: This is not mission-critical, but I'm marking the PR as a draft until this is resolved.

File: data_model/src/lib.rs (hunk 9 above)
Comment (appetrosyan) [reply data_model/src/lib.rs:680]: > we can't use TtryInto as it has too many fields.

what do you mean?

Comment (ales-tsurko) [reply data_model/src/lib.rs:680]: First I thought that it doesn't work for complex structures (with more then 2 fields), but then this: https://jeltef.github.io/derive_more/derive_more/try_into.html#structs

**Review by QuentinI — APPROVED** [review]

File: data_model/src/lib.rs (hunk 9 above)
Comment (appetrosyan) [reply data_model/src/lib.rs:688]: I'm adding a ticket to derive it.

**Review by appetrosyan — APPROVED** [review]

File: tools/parity_scale_decoder/src/generate_map.rs [hunk 11]
```diff
@@ -95,6 +95,7 @@ pub fn generate_map() -> DumpDecodedMap {
         BlockHeaderValue,
         BlockRejectionReason,
         BlockValue,
+        BlockValueWrapper,
```
Comment (Erigara) [inline tools/parity_scale_decoder/src/generate_map.rs:98]: I think this line is useless now, because we hide `BlockValueWrapper`.

**Review by Erigara — APPROVED** [review]

**Review by appetrosyan — APPROVED** [review]

**Pull request merged**